tests/unit/tests_high_five_base.cpp

```diff
@@ -1618,6 +1618,173 @@ TEMPLATE_LIST_TEST_CASE("irregularHyperSlabSelectionWrite", "[template]", std::t
+        subshape[1] += xslice[1] - xslice[0];
+    }
+
+    REQUIRE(subarray.size() == subshape[0]);
+    for (const auto& v: subarray) {
+        REQUIRE(v.size() == subshape[1]);
+    }
+}
+
+template <class S, class Y, class X>
+void check_product_set_values(const S& array,
+                              const S& subarray,
+                              const Y& yslices,
+                              const X& xslices) {
+    size_t il = 0;
+    for (const auto& yslice: yslices) {
+        for (size_t ig = yslice[0]; ig < yslice[1]; ++ig) {
+            size_t jl = 0;
+
+            for (const auto& xslice: xslices) {
+                for (size_t jg = xslice[0]; jg < xslice[1]; ++jg) {
+                    REQUIRE(subarray[il][jl] == array[ig][jg]);
+                    ++jl;
+                }
+            }
+            ++il;
+        }
+    }
+}
+
+template <class S, class Y, class X>
+void check(const S& array, const S& subarray, const Y& yslices, const X& xslices) {
+    check_product_set_shape(subarray, yslices, xslices);
+    check_product_set_values(array, subarray, yslices, xslices);
+};
+
+
+TEST_CASE("productSet") {
+    using Slice = std::array<size_t, 2>;
+    using Slices = std::vector<Slice>;
+    using Point = size_t;
+    using Points = std::vector<size_t>;
+
+    const std::string file_name("h5_test_product_set.h5");
+
+    auto generate = [](size_t n, size_t m, auto f) {
+        auto x = std::vector<std::vector<double>>(n);
+        for (size_t i = 0; i < n; ++i) {
+            x[i] = std::vector<double>(m);
+        }
+
+        for (size_t i = 0; i < n; ++i) {
+            for (size_t j = 0; j < m; ++j) {
+                x[i][j] = f(i, j);
+            }
+        }
+
+        return x;
+    };
+
+    auto array = generate(6, 12, [](size_t i, size_t j) { return double(i) + double(j) * 0.01; });
+
+    auto file = File(file_name, File::Truncate);
+    auto dset = file.createDataSet("dset", array);
+
+    SECTION("rR") {
+        std::vector<std::vector<double>> subarray;
+
+        auto yslice = Slice{1, 3};
+        auto yslices = Slices{yslice};
+        auto xslices = Slices{{0, 1}, {3, 5}};
+
+        dset.select(ProductSet(yslice, xslices)).read(subarray);
+
+        check(array, subarray, yslices, xslices);
+    }
+
+    SECTION("Rr") {
+        std::vector<std::vector<double>> subarray;
+
+        auto yslices = Slices{{0, 1}, {3, 5}};
+        auto xslice = Slice{1, 3};
+        auto xslices = Slices{xslice};
+
+        dset.select(ProductSet(yslices, xslice)).read(subarray);
+
+        check(array, subarray, yslices, xslices);
+    }
+
+    SECTION("RP") {
+        std::vector<std::vector<double>> subarray;
+
+        auto yslices = Slices{{0, 1}, {3, 5}};
+        auto xpoints = Points{2, 4, 5};
+        auto xslices = Slices{{2, 3}, {4, 6}};
+
+        dset.select(ProductSet(yslices, xpoints)).read(subarray);
+
+        check(array, subarray, yslices, xslices);
+    }
+
+    SECTION("pR") {
```

Comment:
What about `RR` and `PP`?

Reply:
I wasn't scared that I'd be missing anything by skipping those, I'll add them.